File: public/install/uninstall.py
#!/usr/bin/env python3
# pages-seo terminal UNINSTALLER (Python flavour).
#
# Downloaded and piped from https://seo.benjaminb.xyz/install/uninstall.py.
# Reverses what install/run.py provisioned: the Cloudflare Pages project,
# the D1 database, and the R2 bucket. Asks before each destructive step.
#
#   curl -fsSL https://seo.benjaminb.xyz/install/uninstall.py | python3
#
# Use --yes to skip every confirmation (CI-style):
#   curl -fsSL …/uninstall.py | python3 - --yes my-project

import json
import os
import re
import shutil
import subprocess
import sys

C = {
    'reset': '\033[0m', 'dim': '\033[2m', 'bold': '\033[1m',
    'cyan':  '\033[36m', 'green': '\033[32m', 'red': '\033[31m', 'yellow':'\033[33m',
}
def say(s):  print(f"{C['cyan']}▸{C['reset']} {C['bold']}{s}{C['reset']}")
def ok(s):   print(f"  {C['green']}✓{C['reset']} {s}")
def warn(s): print(f"  {C['yellow']}!{C['reset']} {s}")
def err(s):  print(f"  {C['red']}✗{C['reset']} {s}")
def die(s):  print(f"{C['red']}✗ {s}{C['reset']}"); sys.exit(1)

# We read confirmations from /dev/tty so the prompt works when this
# script is piped through python3 (stdin is the script body, not the
# terminal). Falls back to plain input() if there's no /dev/tty.
def _tty_io():
    try:
        return open('/dev/tty', 'r'), open('/dev/tty', 'w')
    except OSError:
        return None, None
def ask(prompt, default=''):
    rin, rout = _tty_io()
    label = prompt + (f' [{default}]' if default else '') + ': '
    if not rin:
        try: ans = input(label).strip()
        except EOFError: ans = ''
        return ans or default
    rout.write(label); rout.flush()
    try: ans = rin.readline().strip()
    finally: rin.close(); rout.close()
    return ans or default
def ask_yes_no(prompt, default_no=True):
    suffix = '[y/N]' if default_no else '[Y/n]'
    a = ask(f'{prompt} {suffix}').lower()
    if not a: return not default_no
    return a in ('y', 'yes')

def run(args):
    return subprocess.run(args, capture_output=True, text=True)

def wrangler_installed():
    return shutil.which('wrangler') is not None or run(['wrangler', '--version']).returncode == 0

def is_logged_in():
    r = run(['wrangler', 'whoami'])
    return r.returncode == 0 and 'not authenticated' not in (r.stdout + r.stderr).lower()

# ── discovery ──────────────────────────────────────────────────────
# Look up the live IDs for the project's D1 + R2 so we delete the right
# things. Wrangler's `pages project list --json` includes deployment
# config which carries the bindings; if that fails we fall back to the
# project name + the conventional "<project>-images" R2 name.

def discover(project):
    found = {'pages': False, 'd1_id': None, 'd1_name': None, 'r2_name': None}

    # Pages project existence
    r = run(['wrangler', 'pages', 'project', 'list'])
    out = (r.stdout or '') + (r.stderr or '')
    if re.search(rf'\b{re.escape(project)}\b', out):
        found['pages'] = True

    # D1 by name match
    r = run(['wrangler', 'd1', 'list', '--json'])
    if r.returncode == 0:
        try:
            m = re.search(r'\[[\s\S]*\]', r.stdout or '')
            if m:
                for row in json.loads(m.group(0)):
                    if row.get('name') == project:
                        found['d1_id'] = row.get('uuid') or row.get('database_id')
                        found['d1_name'] = row.get('name')
                        break
        except Exception:
            pass

    # R2 bucket: convention is "<project>-images" (run.py / run.sh / run.js).
    r2_guess = project + '-images'
    r = run(['wrangler', 'r2', 'bucket', 'list'])
    if r.returncode == 0 and re.search(rf'\b{re.escape(r2_guess)}\b', r.stdout or ''):
        found['r2_name'] = r2_guess
    return found

# ── deletions ──────────────────────────────────────────────────────
def delete_pages(project, yes):
    if not yes and not ask_yes_no(f'Delete Pages project "{project}"?'):
        warn('skipped Pages project'); return
    r = run(['wrangler', 'pages', 'project', 'delete', project, '--yes'])
    out = (r.stdout or '') + (r.stderr or '')
    if r.returncode == 0 or 'not found' in out.lower():
        ok(f'deleted Pages project "{project}"')
    else:
        err(f'failed to delete Pages project:\n{out.strip()}')

def delete_d1(d1_id, d1_name, yes):
    if not yes and not ask_yes_no(f'Delete D1 database "{d1_name}" ({d1_id})? This wipes all blog posts.'):
        warn('skipped D1 database'); return
    r = run(['wrangler', 'd1', 'delete', d1_id, '--skip-confirmation'])
    out = (r.stdout or '') + (r.stderr or '')
    if r.returncode == 0 or 'not found' in out.lower():
        ok(f'deleted D1 database "{d1_name}"')
    else:
        err(f'failed to delete D1:\n{out.strip()}')

def delete_r2(r2_name, yes):
    if not yes and not ask_yes_no(f'Delete R2 bucket "{r2_name}"? This wipes all hero images.'):
        warn('skipped R2 bucket'); return
    # Empty the bucket first so the delete doesn't fail on non-empty.
    run(['wrangler', 'r2', 'bucket', 'delete', r2_name, '--force'])  # newer wrangler
    r = run(['wrangler', 'r2', 'bucket', 'delete', r2_name])
    out = (r.stdout or '') + (r.stderr or '')
    if r.returncode == 0 or 'not found' in out.lower() or 'does not exist' in out.lower():
        ok(f'deleted R2 bucket "{r2_name}"')
    else:
        err(f'failed to delete R2 bucket:\n{out.strip()}')

# ── main ───────────────────────────────────────────────────────────
def main():
    print()
    print(f"{C['cyan']}╭──────────────────────────────────────────────╮{C['reset']}")
    print(f"{C['cyan']}│{C['reset']}  {C['bold']}pages-seo · UNINSTALL{C['reset']}                     {C['cyan']}│{C['reset']}")
    print(f"{C['cyan']}│{C['reset']}  {C['dim']}removes the Pages project + D1 + R2{C['reset']}        {C['cyan']}│{C['reset']}")
    print(f"{C['cyan']}╰──────────────────────────────────────────────╯{C['reset']}")
    print()

    # Parse args (yes-flag + optional project slug). When piped via curl
    # … | python3 - --yes my-project, argv[1:] is the trailing args.
    args = sys.argv[1:]
    yes = '--yes' in args or '-y' in args
    args = [a for a in args if a not in ('--yes', '-y')]

    if not wrangler_installed():
        die('wrangler is not installed. Install it first: npm install -g wrangler')
    if not is_logged_in():
        warn('Not logged in to Cloudflare — running `wrangler login`…')
        run(['wrangler', 'login'])
        if not is_logged_in(): die('Login failed.')
    ok('logged in to Cloudflare')

    project = args[0] if args else ask('Project slug to uninstall (the one you used when installing)')
    project = project.strip().lower()
    if not re.match(r'^[a-z][a-z0-9-]{1,32}$', project):
        die('Project slug must be lowercase letters/digits/dashes (e.g. my-blog).')

    print()
    say(f'Looking up resources for "{project}"')
    f = discover(project)

    summary = []
    if f['pages']: summary.append(f'Pages project "{project}"')
    if f['d1_id']: summary.append(f'D1 database "{f["d1_name"]}" ({f["d1_id"]})')
    if f['r2_name']: summary.append(f'R2 bucket "{f["r2_name"]}"')

    if not summary:
        warn('Nothing found to delete for that slug.')
        warn('Common cause: typo in the slug. Re-run with the exact name you gave the installer.')
        return

    print()
    print('  Found:')
    for s in summary: print(f'    · {s}')
    print()
    print(f"  {C['yellow']}This is destructive and cannot be undone.{C['reset']}")
    if not yes:
        if not ask_yes_no('Proceed with deletion?'): die('Aborted.')
    print()

    say('Removing resources')
    if f['pages']:  delete_pages(project, yes)
    if f['d1_id']:  delete_d1(f['d1_id'], f['d1_name'], yes)
    if f['r2_name']: delete_r2(f['r2_name'], yes)

    print()
    print(f"  {C['bold']}{C['green']}Uninstall complete.{C['reset']}")
    print(f"  Your Cloudflare account no longer has any pages-seo resources for \"{project}\".")
    print()

if __name__ == '__main__':
    try: main()
    except KeyboardInterrupt:
        print(); die('Cancelled.')

File: public/install/uninstall.js
#!/usr/bin/env node
// pages-seo terminal UNINSTALLER (Node flavour).
//
//   curl -fsSL https://seo.benjaminb.xyz/install/uninstall.js | node
//   curl -fsSL https://seo.benjaminb.xyz/install/uninstall.js | node - --yes my-project
//
// Removes the Cloudflare Pages project, D1 database, and R2 bucket
// that install/run.js created. Asks before each destructive step
// (unless --yes is passed).

import { spawnSync } from 'node:child_process';
import { createInterface } from 'node:readline';
import { existsSync, openSync } from 'node:fs';
import { ReadStream as TTYReadStream } from 'node:tty';

// When run via `curl … | node`, stdin IS the script body. Re-open
// /dev/tty so prompts read the user, not the script.
let input = process.stdin;
try {
  if (!process.stdin.isTTY && existsSync('/dev/tty')) {
    const fd = openSync('/dev/tty', 'r');
    input = new TTYReadStream(fd);
  }
} catch { /* */ }
const output = process.stdout;

const C = {
  reset: '\x1b[0m', dim: '\x1b[2m', bold: '\x1b[1m',
  cyan:  '\x1b[36m', green: '\x1b[32m', red: '\x1b[31m', yellow: '\x1b[33m',
};
const log  = (...a) => console.log(...a);
const say  = (s) => log(`${C.cyan}▸${C.reset} ${C.bold}${s}${C.reset}`);
const ok   = (s) => log(`  ${C.green}✓${C.reset} ${s}`);
const warn = (s) => log(`  ${C.yellow}!${C.reset} ${s}`);
const err  = (s) => log(`  ${C.red}✗${C.reset} ${s}`);
const die  = (s) => { log(`${C.red}✗ ${s}${C.reset}`); process.exit(1); };

function run(cmd, args) {
  return spawnSync(cmd, args, { encoding: 'utf8' });
}
function runOk(cmd, args) { return run(cmd, args).status === 0; }

const rl = createInterface({ input, output });
function ask(q, def = '') {
  return new Promise((res) => {
    rl.question(`  ${q}${def ? ` [${def}]` : ''}: `, (a) => res((a || '').trim() || def));
  });
}
async function askYesNo(q, defaultNo = true) {
  const ans = (await ask(`${q} ${defaultNo ? '[y/N]' : '[Y/n]'}`)).toLowerCase();
  if (!ans) return !defaultNo;
  return ans === 'y' || ans === 'yes';
}

const ARGS = process.argv.slice(2);
const YES = ARGS.includes('--yes') || ARGS.includes('-y');
const POS = ARGS.filter((a) => a !== '--yes' && a !== '-y');

function banner() {
  log('');
  log(`${C.cyan}╭──────────────────────────────────────────────╮${C.reset}`);
  log(`${C.cyan}│${C.reset}  ${C.bold}pages-seo · UNINSTALL${C.reset}                     ${C.cyan}│${C.reset}`);
  log(`${C.cyan}│${C.reset}  ${C.dim}removes the Pages project + D1 + R2${C.reset}        ${C.cyan}│${C.reset}`);
  log(`${C.cyan}╰──────────────────────────────────────────────╯${C.reset}`);
  log('');
}

function discoverPages(project) {
  const r = run('wrangler', ['pages', 'project', 'list']);
  const out = (r.stdout || '') + (r.stderr || '');
  return new RegExp(`\\b${project.replace(/[.*+?^${}()|[\\]\\\\]/g, '\\$&')}\\b`).test(out);
}
function discoverD1(project) {
  const r = run('wrangler', ['d1', 'list', '--json']);
  if (r.status !== 0) return null;
  try {
    const m = (r.stdout || '').match(/\[[\s\S]*\]/);
    if (!m) return null;
    const rows = JSON.parse(m[0]);
    const hit = rows.find((x) => x && x.name === project);
    if (!hit) return null;
    return { id: hit.uuid || hit.database_id, name: hit.name };
  } catch { return null; }
}
function discoverR2(project) {
  const name = `${project}-images`;
  const r = run('wrangler', ['r2', 'bucket', 'list']);
  if (r.status !== 0) return null;
  const out = (r.stdout || '') + (r.stderr || '');
  return new RegExp(`\\b${name.replace(/[.*+?^${}()|[\\]\\\\]/g, '\\$&')}\\b`).test(out) ? name : null;
}

async function deletePages(project) {
  if (!YES && !(await askYesNo(`Delete Pages project "${project}"?`))) { warn('skipped Pages project'); return; }
  const r = run('wrangler', ['pages', 'project', 'delete', project, '--yes']);
  const out = (r.stdout || '') + (r.stderr || '');
  if (r.status === 0 || /not found/i.test(out)) ok(`deleted Pages project "${project}"`);
  else err(`Pages delete failed: ${out.trim()}`);
}
async function deleteD1(d1) {
  if (!YES && !(await askYesNo(`Delete D1 database "${d1.name}" (${d1.id})? This wipes all blog posts.`))) { warn('skipped D1'); return; }
  const r = run('wrangler', ['d1', 'delete', d1.id, '--skip-confirmation']);
  const out = (r.stdout || '') + (r.stderr || '');
  if (r.status === 0 || /not found/i.test(out)) ok(`deleted D1 database "${d1.name}"`);
  else err(`D1 delete failed: ${out.trim()}`);
}
async function deleteR2(name) {
  if (!YES && !(await askYesNo(`Delete R2 bucket "${name}"? This wipes all hero images.`))) { warn('skipped R2 bucket'); return; }
  let r = run('wrangler', ['r2', 'bucket', 'delete', name, '--force']);
  if (r.status !== 0) r = run('wrangler', ['r2', 'bucket', 'delete', name]);
  const out = (r.stdout || '') + (r.stderr || '');
  if (r.status === 0 || /not found|does not exist/i.test(out)) ok(`deleted R2 bucket "${name}"`);
  else err(`R2 delete failed: ${out.trim()}`);
}

async function main() {
  banner();

  if (!runOk('wrangler', ['--version'])) die('wrangler is not installed. Install it first: npm install -g wrangler');
  if (!runOk('wrangler', ['whoami'])) {
    warn('Not logged in to Cloudflare — running `wrangler login`…');
    run('wrangler', ['login']);
    if (!runOk('wrangler', ['whoami'])) die('Login failed.');
  }
  ok('logged in to Cloudflare');

  let project = (POS[0] || await ask('Project slug to uninstall (the one you used when installing)')).trim().toLowerCase();
  if (!/^[a-z][a-z0-9-]{1,32}$/.test(project)) die('Project slug must be lowercase letters/digits/dashes (e.g. my-blog).');

  log('');
  say(`Looking up resources for "${project}"`);
  const havePages = discoverPages(project);
  const d1 = discoverD1(project);
  const r2 = discoverR2(project);

  const summary = [];
  if (havePages) summary.push(`Pages project "${project}"`);
  if (d1) summary.push(`D1 database "${d1.name}" (${d1.id})`);
  if (r2) summary.push(`R2 bucket "${r2}"`);

  if (!summary.length) {
    warn('Nothing found to delete for that slug.');
    warn('Common cause: typo in the slug. Re-run with the exact name you gave the installer.');
    rl.close(); return;
  }

  log('  Found:');
  for (const s of summary) log(`    · ${s}`);
  log('');
  log(`  ${C.yellow}This is destructive and cannot be undone.${C.reset}`);
  if (!YES && !(await askYesNo('Proceed with deletion?'))) die('Aborted.');
  log('');

  say('Removing resources');
  if (havePages) await deletePages(project);
  if (d1)        await deleteD1(d1);
  if (r2)        await deleteR2(r2);

  log('');
  log(`  ${C.bold}${C.green}Uninstall complete.${C.reset}`);
  log(`  Your Cloudflare account no longer has any pages-seo resources for "${project}".`);
  log('');
  rl.close();
}

main().catch((e) => { err(String(e?.stack || e)); process.exit(1); });
